### DISCUSSION TITLE
Virtual World URIs

### DISCUSSION THREAD

**Discussion by humbletim** [omigroup/omigroup#23] [General]
a few conversation starters:
- what are some benefits of having interoperable cross-world, cross-platform hyperlinks?
- what high-level use cases are we already aware of?
  - links that enable jumping between platform-worlds
  - links to inventory?
  - links to identity / friends?
  - ... ?
- what components would be useful to embed within metaverse hyperlinks?
- how do existing virtual world platforms solve this problem internally and/or externally?
- are there existing URI standards to draw upon?
- ...

**Comment by humbletim** [comment]
(incomplete) list of existing virtual world hyperlink formats.
```
https://hub.link/{AbcPQr}
https://play.decentraland.org/?position={X},{Y}
https://www.cryptovoxels.com/play?coords={lon}W,{lat}S
https://vesta.janusxr.org/{user}/{world}
vrchat://launch?id=wrld_{UUID}
recroom://https://rec.net/room/{RoomName}
steam://rungameid/505700//{world} # anyland
neos:?world=neos-session:///S-{UUID}
hop://login.osgrid.org:80/{Region}/{X}/{Y}/{Z}/
hifi://{placename}/{X,Y,Z}/{X,Y,Z,W}
hifi://{UUID}/{locallyDefinedPaths} # each world resolves its own human-friendly paths into local world coordinates
```
note: i think several of these platforms also accept additional information in world/instance URLs

**Reply by technobaboo** [reply]
I can see 2 main patterns here: world ID/session (sometimes including user), and coordinates. From all this, a URI could look something like `{platform}://{world}/{session}` or `{platform}://{x}/{y}/{z}`

**Reply by AdamFrisby** [reply]
We use sinespace://{world}/{X}/{Y}/{Z} (this works today. X/Y/Z are optional.)

However that is going to be changing to sinespace://{host}/{grid}/{world}/{X}/{Y}/{Z} in a future update.

**Comment by mrmetaverse** [comment]
We discussed this topic to [todays OMI meeting](https://hackmd.io/@mrmetaverse/SJRw4GH9u)

**Comment by gobengo** [comment]
We talked about using [WebFinger](https://datatracker.ietf.org/doc/html/rfc7033) to resolve related resources to a world uri:
* https://datatracker.ietf.org/doc/html/rfc7033

@technobaboo is saying the things that would be useful to 'discover' about a World URI are:
* flat image (e.g. 'icon')
* equirectangular world preview for portals
* pretty printed world name, portal name, world instance

@jsealt also took notes as https://discord.com/channels/770382203782692945/835221348443095070/849698892639240252

**Comment by madjin** [comment]
there's this https://github.com/common-metaverse/urn-namespaces

one use-case can be a way to inspect 3rd party assets that come from other worlds, like a hat that comes from cryptovoxels for instance.

**Reply by menduz** [reply]
Hello, I am the creator of the urn-namespaces initiative. I'd love to get involved in this OMI conversations, and I believe omigroup would be a better home for urn-namespaces if we decide to go in that path. 

I'm in discord if anyone wants to talk menduz#9700

**Reply by robertlong** [reply]
@menduz do you think you'll be around for our weekly meeting tomorrow or next week? I'd like to hear more about this project!

**Reply by menduz** [reply]
When it is scheduled? May I have an invite? `[my last name] at decentraland.org`

**Reply by mrmetaverse** [reply]
@menduz just following up, we meet in the [OMI/Angell ](https://discord.gg/NJtT9grz5E)discord every Wednesday at 5pm UTC as a group, and our working groups meet on other days and times throughout the week. We would love your perspective on the URI efforts. We leverage github as a knowledge repo, so feel free to start new discussions, propose new issues, or reach out to me on Discord!

**Reply by menduz** [reply]
My perspective on this particular topic is that: it is too early to try to standardize anything related to URI schemes (or protocol), every implementation should do what works best for their platform, universe, realm, coordinate system, etc. Iterate and polish their implementation, and talk again once they have found and proof what really works.

For URN, Decentraland is making heavy use of the URNs to identify and resolve assets to servers, contracts and IPFS hashes. It is working _really_ well. Sadly the repository did not receive any new submission from other projects, I assume they didn't hit the wall of interoperability yet. You can take a look at the URN resolvers here: https://github.com/decentraland/urn-resolver/blob/main/src/resolvers.ts#L20-L41.

Here is an example about integrating third party assets to our URNs to get a canonical representation of the external IDs https://github.com/decentraland/adr/blob/main/docs/ADR-42-third-party-assets-integration.md

**Reply by humbletim** [reply]
> My perspective on this particular topic is that: it is too early to try to standardize anything related to URI schemes (or protocol), every implementation should do what works best for their platform, universe, realm, coordinate system, etc. Iterate and polish their implementation, and talk again once they have found and proof what really works.

But is what works best for a platform at the local level... necessarily what would work best for them _at an ecosystem level_?

For example, platforms adopting an opaque URL format (basically containing a primary key pointing to their central database) on one hand make for something potentially easy to maintain as a permalink and easy for users to copy and paste -- but on the other lead to a hyperlink format that is somewhat sterile and inflexible in terms of leveraging network effects.

In the web world, consider how [slugs](https://seo-hacker.com/url-seo-tutorial/) (human friendly, SEO friendly, content relevant paths) have become a de facto standard for how we represent blog entry URLs... generally leading to hyperlinks that:
- are transparent and informative even _on their own_.
- can be _understood by humans_.
- can be readily _discussed with friends_.
- work well inside of _curated lists of hyperlinks_.
- integrate well with _search engines_ (and _search results_).
- provide _useful clues_ as to what a _user can expect_ if choosing to follow a link.

In context of virtual worlds, consider these two approaches:

#### Walled-garden like opaque URLs
**https://opaq.ue/0000-0000-0000-0001**

  - Perhaps easier to implement (all links are effectively owned by a central database, indexed by uuid key).
  - Refers to something on the brand's grid...
    - ... but what? to know what means having to traverse the link.
  - If `opaq.ue` server is down, consumer is dead in the water.
  - If `opaq.ue` returns a 404 page, consumer is also likely dead in the water.

#### Rich URLs with macro-to-micro coordinates
**https://a.universe.name/region/parcel/spawnpoint**

  - Perhaps harder to implement (requires an upfront specification; and revision management if varying later).
  - Refers to a specific location within the "Universe" realm.
    - ... we can see which universe, region, parcel, and spawnpoint (without traversing).
  - If `a.universe.name` server is down,  a user or bug report remains conversationally detailed:
    - "Hey, I'm trying to join a meeting at Region in Universe... can you help me get there?"
  - If `a.universe.name` returns a 404 page, consumer has some systematic recovery options:
    - eg: could work "upwards" (`..`) through the URL path segments.
  - Has built-in SEO potential -- eg: a search engine could index and correlate into nearby results by region
  - Adaptable -- eg: a savvy user could switch their spawnpoint themselves (and autonomously share the modified URL with friends)
  - Lateral mapping potential -- eg: new URLs and hyperlinking relationships can be synthesized from parts

#### TL;DR

We may not need to standardize on an exact URL format here -- but it seems like the time is definitely now to begin discussing overall URL strategies, and in particular, strategies seen from the standpoint of _external_ consumers of our hyperlinks.

**Comment by technobaboo** [comment]
I just want to add a story here so we have context as to why this is important:

Right Now: My friend just saw this very neat hangout world in VR, and when they send it to me on Discord/Matrix/Twitter I just see a link, so I have to now click it to go to a webpage describing the world in more detail. But the problem is, some platforms will open the world in that place (e.g. cryptovoxels) and some will pop up a page showing the info and having a button to launch, which can be slow and is not inline. I now have to actually go to the place to get basic information, or go to an intermediate page to see more about that world.

Ideal: My friend saw this very neat hangout world in VR, and when they send it to me I see the world title, description, a nice image/equirectangular map, and maybe who can access it all in the messaging app or social media. I know all the basics about the world, and can decide if I have the energy, time, proper software, ability, etc. to join them before making any commitment to joining.

Currently there is no standard or even set of expected metadata to send over to messaging apps or such so that the user can see what they're getting into before they go. This discussion is to hash out a standard or set of expected data to send over.

**Reply by josephrocca** [reply]
> Currently there is no standard or even set of expected metadata to send over to messaging apps

As a point of reference, the web has settled on open graph tags (and some custom meta tags for large sites like twitter, fb, etc.) to solve this problem. The meta tags power link preview modules which show the image, title, etc. when pasted in messengers and other communication apps. Here's an example of some of the meta tags for [this article](https://www.theguardian.com/film/2021/nov/11/disney-is-latest-firm-to-announce-metaverse-plans):

```
<meta property="og:url" content="https://www.theguardian.com/film/2021/nov/11/disney-is-latest-firm-to-announce-metaverse-plans"/>
<meta property="og:image:height" content="720"/>
<meta property="og:description" content="Entertainment company plans to ‘connect the physical and digital worlds … allowing for storytelling without boundaries’"/>
<meta property="og:image:width" content="1200"/>
<meta property="og:image" content="https://i.guim.co.uk/img/media/87a451ddbcece3cdcbbc2545b950cad66cbfe93a/762_0_3115_1870/master/3115.jpg?width=1200&height=630&quality=85&auto=format&fit=crop&overlay-align=bottom%2Cleft&overlay-width=100p&overlay-base64=L2ltZy9zdGF0aWMvb3ZlcmxheXMvdGctZGVmYXVsdC5wbmc&enable=upscale&s=04ea3134fad2e314eebc144561656be3"/>
<meta property="og:type" content="article"/>
<meta property="og:title" content="A whole new world: Disney is latest firm to announce metaverse plans"/>
<meta property="og:site_name" content="the Guardian"/>

<meta property="article:author" content="https://www.theguardian.com/profile/danmilmo"/>
<meta property="article:publisher" content="https://www.facebook.com/theguardian"/>
<meta property="article:section" content="Film"/>
<meta property="article:published_time" content="2021-11-11T15:56:56.000Z"/>
<meta property="article:tag" content="Walt Disney Company,Disney+,Media,Film,Technology,Virtual reality,Culture"/>
<meta property="article:modified_time" content="2021-11-12T12:37:40.000Z"/>

<meta property="fb:app_id" content="180444840287"/>

<meta property="al:ios:app_store_id" content="409128287"/>
<meta property="al:ios:app_name" content="The Guardian"/>
<meta property="al:ios:url" content="gnmguardian://film/2021/nov/11/disney-is-latest-firm-to-announce-metaverse-plans?contenttype=Article&source=applinks"/>

<meta name="twitter:dnt" content="on">
<meta name="twitter:app:id:iphone" content="409128287"/>
<meta name="twitter:app:name:googleplay" content="The Guardian"/>
<meta name="twitter:app:name:ipad" content="The Guardian"/>
<meta name="twitter:image" content="https://i.guim.co.uk/img/media/87a451ddbcece3cdcbbc2545b950cad66cbfe93a/762_0_3115_1870/master/3115.jpg?width=1200&height=630&quality=85&auto=format&fit=crop&overlay-align=bottom%2Cleft&overlay-width=100p&overlay-base64=L2ltZy9zdGF0aWMvb3ZlcmxheXMvdGctZGVmYXVsdC5wbmc&s=7f458439821058070b8bcf43d758d510"/>
<meta name="twitter:site" content="@guardian"/>
<meta name="twitter:app:url:ipad" content="gnmguardian://film/2021/nov/11/disney-is-latest-firm-to-announce-metaverse-plans?contenttype=Article&source=twitter"/>
<meta name="twitter:card" content="summary_large_image"/>
<meta name="twitter:app:name:iphone" content="The Guardian"/>
<meta name="twitter:app:id:ipad" content="409128287"/>
<meta name="twitter:app:id:googleplay" content="com.guardian"/>
<meta name="twitter:app:url:googleplay" content="guardian://www.theguardian.com/film/2021/nov/11/disney-is-latest-firm-to-announce-metaverse-plans"/>
<meta name="twitter:app:url:iphone" content="gnmguardian://film/2021/nov/11/disney-is-latest-firm-to-announce-metaverse-plans?contenttype=Article&source=twitter"/>
```

**Comment by jamesjackson** [comment]
Hi, James here from Open AR Cloud (OARC). I don't mean to derail things but could there also be a case for Real World URIs ?

**Reply by humbletim** [reply]
yes definitely and i think it makes sense to anticipate having some kind of common baseline that can be extended for different spatial computing contexts, both earthly and virtual worlds. have you come across any examples of AR apps using (anything similar to) URIs for exchanging location information yet?

**Comment by jfenton** [comment]
From a "launching into that world" context, I noticed VRChat asks you to install an extra "VRChat Helper" which is how it hooks links like: https://vrchat.com/launch?worldId=wrld_6caf5200-70e1-46c2-b043-e3c4abe69e0f&ref=vrchat.com
(I'm not sure why they kept it separate to the main client)

For discussion, perhaps a good initiative for OMI would be to create an OMI Helper which subsumes that kind of functionality in a world-agnostic way? 

So the intent would be to give all Virtual World Platforms ("VWPs") the ability to deep-link to *any* target VWP, without having a priori knowledge of it. Previews etc. are separate, that would be dealt with via WebFinger as discussed, this would only be concerned with the lifecycle of transitioning you to another VWP in a standard VWP-agnostic way.

At first I was thinking of an omi://<world>//... protocol handler at the OS level, particularly for the aesthetics, but I don't think there's a way for the source VWP to get data back (for preview, dependency check, launch status etc.). Anyone know if that's possible? It does also seem more standards-friendly to me if so.

Another method is a basic localhost HTTP(S) protocol that allows the source VWP to query and/or launch into that target VWP's world by talking to a local agent.

Ideas:

- Standard HTTP localhost:XXX protocol on a registered IANA port, for all locally installed engines (and local browsers) to talk to
- Helper could:
- Check whether dependencies are met (VWP installed)
- Bring target VWP to the foreground if already running, and communicate target world to an already-running background VWP
- Or launch into the target VWP if not, using that targets method of passing in the target world URL
- Could potentially suspend the source VWP once transitioned using Nyrna (https://github.com/Merrit/nyrna)
- If not suspend, can provide feedback status to the source VWP on target VWP launch success so it can quiesce or exit
- Can direct you to Installation page for the target VWP if not installed
- Ships as a locally installed native OMI Helper App, which VWPs can optionally sidecar package
- Tiny footprint, minimal core, few app updates, dynamic updates to registered world protocols pulled from OMI periodically (where it gets it's knowledge of how to launch the target world, deps etc.)
- Could provide a beach head to expand into Constellation (future) for ease-of-installation of target VWPs/dependencies

Note that the agent could bridge between the standard local OMI protocol adopted by the source VWP, and a target VWP's potentially proprietary method - it wouldn't necessarily need target VWP support. I would hope we could get the larger "source" VWPs to adopt initially (for example, approaching VRChat and asking them to migrate their helper functionality us would be a great first start).

In similar fashion to IPSME brokering, this approach allows us to bridge between worlds that may already have existing deep-linking abilities, but would perhaps be reticent (or slow) to adopt an OMI standard just yet. Also, without a local app (or embedded OMI library in a VWP) we can't nail the launching ability for anything other than web-based worlds.

Comments welcome.

**Reply by humbletim** [reply]
there is definitely shared interest around having ways to load virtual world URIs locally on demand -- i'm going to add to this week's agenda for further discussion. :)

also a while back when brainstorming with @fire and others about OS-level protocol handlers -- one stumbling block was how a courteous platform app could (even in theory) "yield the floor" to a sibling platform. Nyrna seems super interesting on that front... have you experimented at all with how VR apps react to be being hibernated/revived that way?

**Reply by mrmetaverse** [reply]
I think this is worth discussion and experimentation. I wonder if this ties in in any way with the OMI-chan project as well, at least further down the roadmap. Helpers are huge value-adds if they are done right.

**Reply by jfenton** [reply]
> also a while back when brainstorming with @fire and others about OS-level protocol handlers -- one stumbling block was how a courteous platform app could (even in theory) "yield the floor" to a sibling platform. Nyrna seems super interesting on that front... have you experimented at all with how VR apps react to be being hibernated/revived that way?

So looking under the hood, Nyrna's just using the win32 suspend process APIs - likely no different to Task Manager > Suspend Process. To answer your question, I suspect "violently". 😂 Maybe no different to alt-tabbing? Can multiple apps bind to the headset/controller telemetry at present? Obviously a well-defined protocol for transitioning the user correctly is far better, otherwise you'd be in two places at once. Which may be uncomfortable.

**Comment by technobaboo** [comment]
I think that the app to app communication is absolutely what should happen,
but using a web server and dedicated manager is a bit clunky and can
instead be replaced with a Stellar Mesh (now known as Constellation)
communication layer using more OS-specific means to simplify and speed up
communication. See https://github.com/omigroup/OMI/discussions/73 for a
discussion on a framework for this that is similar, and check the discord
for #omi-constellation and see if that meets all the needs mentioned here.

**Reply by jfenton** [reply]
Firstly, thanks for your comments Nova! I'm absolutely supportive of and across Constellation (and love the new name, as we all do!) - just been lurking due to time constraints. My only concern is with tying Constellation too much to OS primitives (that only exist on Linux presently), as I personally feel this will slow adoption and integration for other platforms. As you mentioned, it's hard to identify real-world useable integrations right now because the existing VWPs are closed boxes. To me, that means making adoption as easy as possible (and compelling enough to adopt) is imperative, and across all of their supported platforms, target OSs and versions.

Perhaps the above Helper becomes the middleware that leverages those OS primitives by proxy, and/or shims equivalents where they don't exist? If OS primitives don't exist pervasively, there is an argument to be made that we should come up a layer, and use something like ZeroMQ or some other cross-platform library to provide our IPC/pub-sub (is D-Bus providing us with more functionality than this?). That way we could ship something embeddable with minimal dependencies for the user.

That all being said - re:a bit clunky, I get ya. My preference would absolutely be a pure protocol approach (and also in keeping with OMIs present goal of not building software). I'm not convinced it's possible at present, because the primitives we need likely don't exist across all targets (as always, happy to be proven wrong on this), but also because a protocol-only approach pushes a larger burden onto the implementors, which again may neuter adoption (and we are already up against the question of why do they want to allow people to leave their cozy walled garden). 

So I suspect, either way, we end up delivering some software off of our standards... if it's a reference implementation / library that makes the OMI protocol more consumable that's a good thing IMHO. There's a caveat that it's harder for us to evolve an embedded library once released. An agent/service approach could become an important beach head for future OMI innovation (and lerping into Constellation, ultimately). It needs to be easy to adopt, and deliver enough value to the VWPs to cause them to adopt right now, whilst these worlds are evolving rapidly.

Removing a burden/overhead from VWPs in supporting myriad Virtual World URIs seems like one of those value props, especially as somebody needs to solve the "oh I don't have that VWP installed" problem. In danger of drifting off-topic, but it's outside of OMIs remit at present to solve that last part, it needs a Steam-esque (but open) approach IMHO. Is that something Constellation will take on as well?

**Comment by technobaboo** [comment]
absolutely there should be a nice middleware to manage the OS-specific
stuff, or at least a common standard. See the constellation discussion for
details and this stuff should mostly go there instead. Constellation is
meant to also deal with containerization/installation.

Virtual world URIs/metadata is still useful even without Constellation, so
having that through Webfinger would be a help for embedding it in sites,
sharing links, etc.

**Comment by humbletim** [comment]
notes from [discord: 2021.08.21 omi-virtual-world-metadata hack-together](https://discord.com/channels/770382203782692945/878654908549713970):

> rough meeting notes:
> * atomic user transfers?
>     - (assuming there's a way to teleport) how to actually "hand off" a user between immersion A and immersion B?
> * resident virtual world overlay / local coordination service?
>     - eg: steam "void" space?
> * discord rich presence?
>     - eg: to offer ways for others to teleport to you
> * some kind of neutral spatial lobby?
>     - junction point for traversing between worlds
> * webfinger / metadata system for validating "teleport" offers?
>     - eg: in response to invites / to make sure they succeed
> * steam, itch, app ecosystems?
> * sounds like we need more research into DID
> [1:23 PM]
> * "hyperminimal presence" (like sip?)
>     - possible implementations?
>     - libp2p
>         - perhaps identity as a DID; maybe ActivityPub over DHT?
>         - (does DID already provide a method for the data layer?)
>         - simple pub/sub?
>     - extensible by WASM plugins? (maybe for later / v2)
>     - what is actually "seeding" the presence information?
>     - is presence specific to relationship, public, both?
> 
> Mauve — 08/21/2021
> Rust useful for WASM. 
> 
> iFire — 08/21/2021
> https://github.com/libp2p/cpp-libp2p useful for WASM 
>

**Comment by humbletim** [comment]
minimum viability thoughts from @fire (see [discord: 2021.08.21 omi-virtual-world-metadata hack-together](https://discord.com/channels/770382203782692945/878654908549713970)):

>  Presence revised design
> 
> * Able to be a tray item and a graphical window.
> * Able to have new friends.
> * Update my current presence as a status message
> * Notification is a visually distinct status that reflects presence of friends.

**Comment by humbletim** [comment]
a resource / reference point offered by A. on discord:
> URI & URL's for OpenSIM
> http://forums.openuru.org/viewtopic.php?t=325
